File: ERP-INVESTIGACION/articulos-backend/src/main/java/com/pe/articulos/modules/permissions/service/PermissionService.java
package com.pe.articulos.modules.permissions.service;

import java.util.List;

import org.springframework.data.domain.Pageable;

import com.pe.articulos.modules.permissions.dto.PermissionRequest;
import com.pe.articulos.modules.permissions.dto.PermissionResponse;
import com.pe.articulos.core.shared.dto.PageResponse;

public interface PermissionService {

    // ========== CRUD BÁSICO ==========

    PermissionResponse createPermission(PermissionRequest request);

    PermissionResponse updatePermission(Long id, PermissionRequest request);

    void deletePermission(Long id);

    // ========== CONSULTAS INDIVIDUALES ==========

    PermissionResponse getPermissionById(Long id);

    PermissionResponse getPermissionByName(String name);

    // ========== LISTADOS CON PAGINACIÓN ==========

    PageResponse<PermissionResponse> getAllPermissions(Pageable pageable);

    PageResponse<PermissionResponse> getActivePermissions(Pageable pageable);

    PageResponse<PermissionResponse> getPermissionsByModule(String module, Pageable pageable);

    PageResponse<PermissionResponse> getActivePermissionsByModule(String module, Pageable pageable);

    PageResponse<PermissionResponse> searchPermissions(String search, String target, Pageable pageable);

    PageResponse<PermissionResponse> searchByFilter(String query, String type, Pageable pageable);

    // ========== LISTADOS SIN PAGINACIÓN (para selects/dropdowns) ==========

    List<PermissionResponse> getAllPermissionsAsList();

    List<PermissionResponse> getActivePermissionsAsList();

    List<PermissionResponse> getPermissionsByModuleAsList(String module);

    List<PermissionResponse> getActivePermissionsByModuleAsList(String module);

    List<PermissionResponse> searchPermissionsAsList(String search);

    // ========== MÓDULOS ==========

    List<String> getAllModules();

    // ========== UTILIDADES ==========

    boolean existsByName(String name);

}

File: ERP-INVESTIGACION/articulos-frontend/src/app/features/configuraciones/services/permission.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { ApiResponse, PageResponse } from '../../../shared/modals/response.model';
import { Permission, PermissionRequest } from '../models/permission.model';

@Injectable({
  providedIn: 'root',
})
export class PermissionService {
  private readonly apiUrl = `${environment.apiUrl}/permissions`;

  constructor(private http: HttpClient) { }

  // ========================================
  // CRUD BÁSICO
  // ========================================

  createPermission(request: PermissionRequest): Observable<ApiResponse<Permission>> {
    return this.http.post<ApiResponse<Permission>>(this.apiUrl, request);
  }

  updatePermission(id: number, request: PermissionRequest): Observable<ApiResponse<Permission>> {
    return this.http.put<ApiResponse<Permission>>(`${this.apiUrl}/${id}`, request);
  }

  deletePermission(id: number): Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(`${this.apiUrl}/${id}`);
  }

  getPermissionById(id: number): Observable<ApiResponse<Permission>> {
    return this.http.get<ApiResponse<Permission>>(`${this.apiUrl}/${id}`);
  }

  getPermissionByName(name: string): Observable<ApiResponse<Permission>> {
    return this.http.get<ApiResponse<Permission>>(`${this.apiUrl}/name/${name}`);
  }

  // ========================================
  // LISTADOS CON PAGINACIÓN
  // ========================================

  private getPaginationParams(page: number, size: number, sort: string): HttpParams {
    return new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString())
      .set('sort', sort);
  }

  getAllPermissions(
    page = 0,
    size = 10,
    sort = 'name'
  ): Observable<ApiResponse<PageResponse<Permission>>> {
    const params = this.getPaginationParams(page, size, sort);
    return this.http.get<ApiResponse<PageResponse<Permission>>>(this.apiUrl, { params });
  }

  getActivePermissions(
    page = 0,
    size = 20,
    sort = 'name'
  ): Observable<ApiResponse<PageResponse<Permission>>> {
    const params = this.getPaginationParams(page, size, sort);
    return this.http.get<ApiResponse<PageResponse<Permission>>>(`${this.apiUrl}/active`, {
      params,
    });
  }

  searchByFilter(
    query: string,
    type: string = 'ALL',
    page: number = 0,
    size: number = 10
  ): Observable<ApiResponse<PageResponse<Permission>>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString())
      .set('sort', 'name')
      .set('query', query)
      .set('type', type);

    return this.http.get<ApiResponse<PageResponse<Permission>>>(`${this.apiUrl}/filter`, {
      params,
    });
  }

  // ========================================
  // LISTADOS SIN PAGINACIÓN (Dropdowns)
  // ========================================

  getAllPermissionsList(): Observable<ApiResponse<Permission[]>> {
    return this.http.get<ApiResponse<Permission[]>>(`${this.apiUrl}/list`);
  }

  getActivePermissionsList(): Observable<ApiResponse<Permission[]>> {
    return this.http.get<ApiResponse<Permission[]>>(`${this.apiUrl}/active/list`);
  }

  // ========================================
  // OPERACIONES POR MÓDULO
  // ========================================

  getPermissionsByModule(
    module: string,
    page = 0,
    size = 10
  ): Observable<ApiResponse<PageResponse<Permission>>> {
    const params = this.getPaginationParams(page, size, 'name');
    return this.http.get<ApiResponse<PageResponse<Permission>>>(`${this.apiUrl}/module/${module}`, {
      params,
    });
  }

  getAllModules(): Observable<ApiResponse<string[]>> {
    return this.http.get<ApiResponse<string[]>>(`${this.apiUrl}/modules`);
  }

  // ========================================
  // UTILIDADES
  // ========================================

  checkPermissionExists(name: string): Observable<ApiResponse<boolean>> {
    return this.http.get<ApiResponse<boolean>>(`${this.apiUrl}/exists/${name}`);
  }
}
